### PR TITLE
refactor load_dataset() functions

### PR DESCRIPTION
**Pull request by lcadalzo**

### CODE REVIEW

**Comment by lcadalzo** [comment]
I'll use this PR thread to share some notes I took after reviewing `datasets.py` and `adversarial_datatsets.py` to capture the range of dataset params we're handling. I ran a script to search every config json in `scenario_configs/` as well as `gard-evaluations` and it outputted the following unique keys, which I've organized into buckets. I've also added a line for each key about where/how it's handled by Armory currently (i.e. in 0.16.0)

```
Required:
'name'
	- popped in config_loading, used to import dataset_fn
'module'
	- "" 
'batch_size'
	- popped in config_loading, passed to dataset_fn
	

Not required, but always used. Set with a default if not present:
'framework'
	- popped in config_loading, passed to dataset_fn
'eval_split'
	- grabbed in scenario.py, passed to config_loading
'train_split'
	- ""
'shuffle_files'
	- gets added to **kwargs that are passed to dataset_fn in config_loading


Filtering:
'index'
	- gets added to **kwargs that are passed to dataset_fn in config_loading
'class_ids'
	- ""


Dataset-specific kwargs:
'modality'
	- ""
'pad_data'
	- ""
'coco_format'
	- ""
'max_frames'
	- ""

```

**Comment by lcadalzo** [comment]
There's a tangled web of places and ways in which we use different dataset params. My hope was to make this more straightforward by having a single `load_dataset()` function that parses the config in a way that's more transparent, and then prepares the call to the actual loader. A few other minor things this PR is doing

- setting `num_eval_batches` scenario attribute to 1 when `--check`. Previously it would be `None`
- set `shuffle_elements` to whatever `shuffle_files` is
- checking whether a default registered `preprocessor` can be found when `preprocessor` is set to `None`, given the dataset `name`

File: armory/scenarios/scenario.py [hunk 1]
```diff
@@ -200,16 +201,36 @@ def load_attack(self):
...
             num_batches=self.num_eval_batches,
-            check_run=self.check_run,
+            epochs=1,
+            split=split,
+            framework=framework,
+            preprocessor_name=preprocessor_name,
+            preprocessor_kwargs=preprocessor_kwargs,
             shuffle_files=False,
+            index=index,
+            class_ids=class_ids,
```
Comment (davidslater) [inline armory/scenarios/scenario.py]: Since we're already modifying the scenario, let's update the config structure as well. I recommend having mirrored dataset fields at the top level, `"test_dataset"` and `"train_dataset"`. Then we could remove `"eval_split"` and `"train_split"`, and just have `"split"` in each of them.

Then this whole section could be simplified to something like:
```python
dataset_config = copy.deepcopy(self.config["test_dataset"])
name = dataset_config.get("name")
log.info(f"Loading test dataset {name}...")
if dataset_config.get("epochs") != 1:
    raise ValueError("epochs must be 1 for test dataset")
if "shuffle_files" not in dataset_config:
    dataset_config["shuffle_files"] = False
dataset_config["split"] = dataset_config.get("split", eval_split_default)
self.test_dataset = config_load.load_dataset(**dataset_config)
```
and similarly for the train loader.

File: armory/datasets/preprocessing.py [hunk 2]
```diff
@@ -100,3 +107,25 @@ def video_to_canon(
...
+        )
+
+
+def carla_over_obj_det_dev_label(x, y_object, y_patch_metadata):
+    # convert TF format to PyTorch format of [x1, y1, x2, y2]
+    height, width = x.shape[1:3]  # TODO: better way to do this?
+    converted_boxes = tf.gather(y_object["boxes"], [1, 0, 3, 2], axis=1)
+    converted_boxes *= [width, height, width, height]
+    y_object["boxes"] = converted_boxes
+    return y_object, y_patch_metadata
```
Comment (davidslater) [inline armory/datasets/preprocessing.py]: This is really hard to follow. Maybe if you start with describing the TF format, then it will make more sense?

File: armory/datasets/config_load.py [hunk 3]
```diff
@@ -49,10 +48,16 @@ def actual_loader(
...
         preprocessor = None
+        if name in preprocessing.list_registered():
+            preprocessor = preprocessing.get(name)
     else:
         preprocessor = preprocessing.get(preprocessor_name)
 
-    return generator.ArmoryDataGenerator(
+    if preprocessor_kwargs is None:
+        preprocessor_kwargs = {}
+    preprocessing_fn = lambda x: preprocessor(x, **preprocessor_kwargs)
```
Comment (davidslater) [inline armory/datasets/config_load.py]: if there aren't any preprocessor_kwargs, we should probably just set `preprocessing_fn = preprocessor`. It will make debugging a lot easier if there are errors in the preprocessor, I think.

File: armory/scenarios/scenario.py (hunk 1 above)
Comment (lcadalzo) [reply armory/scenarios/scenario.py]: I had a similar thought, of renaming `eval_split` to `split` so this could be done. I had been thinking of moving train split to the `"fit_kwargs"` field, but I didn't like having a dataset param under the `"model"` config. I also don't love changing the top-level config structure but I do think in theory it could make sense here, so perhaps it's worth it

An alternative would be to keep the field names and pop them, but that gets closer to what we're already doing currently
```
datset_config["split"] = datset_config.pop("eval_split", None)
```

I think one minor issue with the snippet you posted is that it doesn't account for `version` or `num_batches=self.num_eval_batches`

Comment (davidslater) [reply armory/scenarios/scenario.py]: I think for version, we just make people specify the version explicitly as a key-value pair, and not have it as part of name. For num_eval_batches, you're right.

Comment (davidslater) [reply armory/scenarios/scenario.py]: Yeah, I don't like having a dataset param under the model config. Definitely gets messy that way.

Comment (lcadalzo) [reply armory/scenarios/scenario.py]: added a new commit to address this

File: armory/datasets/config_load.py (hunk 3 above)
Comment (lcadalzo) [reply armory/datasets/config_load.py]: agreed, added a commit to address

**Comment by lcadalzo** [comment]
I haven't changed the config structure to include `"train_dataset"` and `"test_dataset"` and remove `"dataset"`. But otherwise this is ready for re-review

One question I'm thinking about is whether we want to try and be backwards-compatible with the current/would-become-old config structure, but I think that might add some ugliness to our code

**Comment by davidslater** [comment]
Given that Matt is reworking the configs anyway, it's probably easier and better to not be backwards compatible.

**Comment by davidslater** [comment]
Perhaps a better structure for the config is just to add key/value pairs to the dataset section, so that everything is in the same place.

I'm going to work on merging this in, and then merging into the tfdsv4 branch on main armory, so that we can continue to iterate quickly.

**Review by davidslater — APPROVED** [review]
LGTM
